src/programs/shoutout-perms.ts

```diff
@@ -0,0 +1,177 @@
+import { Command, CommandHandler, DiscordEvent } from "../event-distribution";
+import {
+  ChannelType,
+  ChatInputCommandInteraction,
+  Guild,
+  Snowflake,
+  OverwriteType,
+  TextChannel,
+  PermissionOverwrites,
+  Message,
+  User,
+} from "discord.js";
+import bot from "..";
+import { ChatNames } from "../collections/chat-names";
+import { createYesBotLogger } from "../log";
+import { Timer } from "@prisma/client";
+import { TimerService } from "./timer/timer.service";
+
+interface ShoutoutPermsData {
+  channelId: Snowflake;
+  userId: Snowflake;
+}
+
+const shoutoutPermsIdentifier = "shoutoutPermsRemove";
+const shoutoutTimers: Record<Snowflake, string> = {};
+
+const getSendPerms = (channel: TextChannel, user: User) => {
+  return channel.permissionOverwrites.cache
+    .filter((c) => c.type === OverwriteType.Member && c.id === user.id)
+    .first();
+};
+
+const getShoutoutsChannel = (guild: Guild) => {
+  return guild.channels.cache
+    .filter(
+      (c) => c.isTextBased() && (c as TextChannel).name === ChatNames.SHOUTOUTS
+    )
+    .first();
+};
+
+const deletePerms = async (
+  userId: Snowflake,
+  sendPerms: PermissionOverwrites
+) => {
+  const timerId = shoutoutTimers[userId];
+  await sendPerms.delete();
+
+  delete shoutoutTimers[userId];
+  if (timerId !== null) TimerService.cancelTimer(timerId);
+};
+
+@Command({
+  event: DiscordEvent.SLASH_COMMAND,
+  root: "shoutout-perms",
+  description: "Manage who can post in the #shoutout channel!",
+  subCommand: "toggle",
+})
+class ShoutoutPermsToggleCommand
+  implements CommandHandler<DiscordEvent.SLASH_COMMAND>
+{
+  async handle(interaction: ChatInputCommandInteraction): Promise<void> {
+    if (!interaction.guild) return;
+    const channel = interaction.channel;
+
+    if (channel?.type !== ChannelType.GuildText) return;
+    if (!channel.name.startsWith("shoutout-")) {
+      await interaction.reply({
+        content: "Can not send this here!",
+        ephemeral: true,
+      });
+      return;
+    }
+
+    // Get handle of the user
+    const user = interaction.client.users.resolve(channel.name.split("-")[1]);
+    if (user === null) {
+      await interaction.reply({
+        content: "Couldn't find the member!",
+        ephemeral: true,
+      });
+      return;
+    }
+
+    // Get the shoutouts channel
+    const shoutouts = getShoutoutsChannel(interaction.guild);
+
+    // TODO: there must be a way to type this in a more typescript-ish way
+    if (shoutouts === null || shoutouts?.type !== ChannelType.GuildText) {
+      await interaction.reply({
+        content: `Couldn't find the <#${ChatNames.SHOUTOUTS}> channel!`,
+        ephemeral: true,
+      });
+      return;
+    }
+
+    // Check whether there's an instance.
+    // If there is an existing perm overwrite, remove it
+    const sendPerms = getSendPerms(shoutouts, user);
+    if (sendPerms) {
+      deletePerms(user.id, sendPerms);
```

Comment:
returned `Promise`s *must* be awaited; since Node 16 any errors that comes up in a Promise that doesn't have either a `.catch` or an `await` inside a `try/catch` will terminate the process. All handlers are wrapped in `try/catch` somewhere deep in the "core" of YesBot but for that to work, `await` is needed.
```suggestion
      await deletePerms(user.id, sendPerms);
```